accepted/single-file/design.md

```diff
@@ -0,0 +1,279 @@
+`?/il.dll`; `/tmp/.net/e53xf3/r2r.dll`; `/usr/local/nuget/dep.dll`
+
+## Testing
+
+* Unit Tests to achieve good coverage on apps using managed, ready-to-run, native code 
+* Tests to verify that framework-dependent and self-contained apps can be published as single file.
+* Tests for ensure that every app model template supported by .NET Core (console, wpf, winforms, web, etc.) can be published as a single file.
+* A subset of CoreCLR Tests
+* End-to-End testing on real world apps such as Roslyn, MusicStore
+* *Measurements*: Publish size and run-time (first run, subsequent runs) for HelloWorld, Roslyn and MusicStore.
+
+## Further Work
+
+#### Bundler Optimizations
+
+Since all the files of an app published as a single-file live together, we can perform the following optimizations
+
+- R2R compile the app and all of its dependent assemblies in a single version-bubble
+- Investigate whether collectively signing the files in an assembly saves space for certificates.
+
+#### APIs to access Bundled files
+
+The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+
+- Reading additional files packaged into the app (ex: data files).
+- Open an assembly for reflection/inspection
+- Load plugins built as single-file class-libs using existing Loader APIs.
+
+In facilitate this usage, we can add an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+
+```C#
+// Open a file embedded in the bundle built for the specified assembly 
+namespace System.Runtime.Loader
+{
+    public partial class Bundle
+    {
+        public static System.IO.Stream GetFileStream(Assembly assembly, string name);
+    }
+}
+```
+
+We can also provide an abstraction that abstracts away the physical location of a file (bundle or disk). For example, add a variant of  `GetFileStream` API that looks for a file in the bundle, and if not found, falls back to disk-lookup.
+
+#### Single-file Plugins
+
+The above design should be extended to seamlessly support single-file publish for plugins.
+
+- The bundler tool will mostly work as-is, regardless of whether we publish an application or class-lib. The binary blob with dependencies can be appended to both native and managed binaries.
+- For host/runtime support, the options are:
+  - Implement plugins using existing infrastructure. For example: Take control of assembly/native binary loads via existing `AssemblyLoadContext` callbacks and events. Extract the files embedded within the single-file plugin using the `GetFileStream()` API and load them on demand.
+  - Have new API to load a single-file plugin, for example: `AssemblyLoadContext.LoadWithEmbeddedDependencies()`.
```

Comment:
@ayende this is a good suggestion.  Can you describe a potential scenario?

Reply:
Consider the case of resources that are bundled into the app. In our case, we have a web app that has a lot of files (js, png, css, etc).
We actually have three separate options here right now. During dev, we might be working on the raw files, the compiled files or during release, loading them from a zip file.

Being able to tell how we are running would be helpful because if `IsBundled == false`, I need to probe a whole different location to see where the file I need to serve it located.

Reply:
Good feedback, and thanks for the scenario details.

Reply:
OK, thanks for the details @ayende. For this situation, I think it would be better served by providing  an abstraction that handles bundles and files seamlessly, as proposed [here](https://github.com/dotnet/designs/pull/52/files#diff-28dba553f5f12520afc2d90d1587ef56R190).

But I think it is also reasonable to expose an API like the following, to support custom scenarios.
```C#
namespace System.Runtime.Loader
{
    public partial class Bundle
    {
        public static bool IsBundle(Assembly assembly);
    }
}
```